### PR TITLE
New functions for the PubChem PUG-REST API

### PR DESCRIPTION
**Pull request by stitam**
The PR is related to issues #206 and #216.

I originally created two new functions. `pc_pugrest()` gave a lot of control over the PUG-REST API and `pc_validate()` validated all inputs to `pc_pugrest()`. The aim was to offer functions for more advanced users to assemble their own requests. However, `pc_validate` became too complex and very difficult to test, and so I decided to just remove both functions. Instead, I created the function `get_sid()` for issue #216 and will edit `get_cid()` for issue #206 after @Aariq updated it according to #193 so there is no conflict. 

PR task list:
- [x] Update NEWS
- [x] Add tests (if appropriate)
- [x] Update documentation with `devtools::document()`
- [x] Check package passed

### CODE REVIEW

**Comment by stitam** [comment]
The formula in #206 didn't work because the query is sometimes asynchronous and returns a listkey which has to be queried in a subsequent request, similarly to ChemSpider. I restructured `get_cid()` to be able handle this situation as well. I tried to implement all query routes that use PUG-REST API to return cids. https://pubchemdocs.ncbi.nlm.nih.gov/pug-rest. For #216 I created new the function `get_sid()`.

File: R/pubchem.R [hunk 1]
```diff
@@ -1,19 +1,56 @@
...
+#' "{\code{fastidentity} | \code{fastsimilarity_2d} | \code{fastsimilarity_3d} |
+#' \code{fastsubstructure} | \code{fastsuperstructure}}/{\code{smiles} |
+#' \code{smarts} | \code{inchi} | \code{sdf} | \code{cid}}", e.g.
+#' \code{from = "fastidentity/smiles"}.
+#' @details \code{<source id>} is any valid PubChem Data Source ID. When
+#' \code{from = "sourceid/<source id>"}, the query is the ID of the substance in
+#' the depositor's database.
+#' @details If \code{from = "sourceall"} the query is one or more valid Pubchem
+#' depositor names. Depositor names are not case sensitive, but are sensitive to
+#' spaces, and special characters may need to be escaped, such as "&" should be
```
Comment (Aariq) [inline R/pubchem.R:47]: Is this something that could be handled by adding a call to `URLencode()`? If it's not an easy fix, then I'm fine with just documenting it---sounds like this is an edge use case anyway.

File: R/pubchem.R [hunk 2]
```diff
@@ -45,102 +83,164 @@
...
+#' get_cid("56-40-6", from = "xref/rn")
+#' # similarity
+#' get_cid(5564, from = "similarity/cid")
+#' get_cid("CCO", from = "similarity/smiles")
+#' # from SID
+#' get_cid("126534046", from = "sid", domain = "substance")
+#' # sourceid
+#' get_cid("VCC957895", from = "sourceid/23706", domain = "substance")
+#' # sourceall
+#' get_cid("Optopharma Ltd", from = "sourceall", domain = "substance")
```
Comment (Aariq) [inline R/pubchem.R:104]: This gives `Bad Request (HTTP 400)` and returns NA for the cid.

Comment (Aariq) [reply R/pubchem.R:104]: but this works:

```
get_cid(URLencode("Optopharma Ltd"), from = "sourceall", domain = "substance")
```

File: R/pubchem.R [hunk 3]
```diff
@@ -45,102 +83,164 @@
...
+      Sys.sleep(rgamma(1, shape = 15, scale = 1 / 10))
+      if (from == "inchi") {
+        qurl <- paste("https://pubchem.ncbi.nlm.nih.gov/rest/pug",
+                      domain, from, "cids", "json", sep = "/")
+        res <- httr::POST(qurl, body = paste0("inchi=", query),
+                          user_agent("webchem"), handle = handle(""))
+      }
+      else {
+        res <- httr::POST(qurl, user_agent("webchem"),
+                          handle = handle(""))
```
Comment (Aariq) [inline R/pubchem.R:195]: add user_agent() ?

File: R/pubchem.R [hunk 4]
```diff
@@ -45,102 +83,164 @@
...
+      }
+      Sys.sleep(rgamma(1, shape = 15, scale = 1 / 10))
+      if (from == "inchi") {
+        qurl <- paste("https://pubchem.ncbi.nlm.nih.gov/rest/pug",
+                      domain, from, "cids", "json", sep = "/")
+        res <- httr::POST(qurl, body = paste0("inchi=", query),
+                          user_agent("webchem"), handle = handle(""))
+      }
+      else {
+        res <- httr::POST(qurl, user_agent("webchem"),
```
Comment (Aariq) [inline R/pubchem.R:194]:
```suggestion
        res <- httr::POST(URLencode(qurl), user_agent("webchem"),
```

File: R/pubchem.R [hunk 5]
```diff
@@ -45,102 +83,164 @@
...
+      }
+      if (res$status_code != 200) {
+        if (res$status_code == 202) {
+          cont <- httr::content(res, type = "text", encoding = "UTF-8")
+          listkey <- jsonlite::fromJSON(cont)$Waiting$ListKey
+          qurl <- paste("https://pubchem.ncbi.nlm.nih.gov/rest/pug/", domain,
+                        "listkey", listkey, "cids", "json", sep = "/")
+          while (res$status_code == 202) {
+            Sys.sleep(5 + rgamma(1, shape = 15, scale = 1 / 10))
+            res <- httr::POST(qurl, user_agent("webchem"), handle = handle(""))
```
Comment (Aariq) [inline R/pubchem.R:205]:
```suggestion
            res <- httr::POST(URLencode(qurl), user_agent("webchem"), handle = handle(""))
```

File: R/pubchem.R (hunk 1 above)
Comment (stitam) [reply R/pubchem.R:47]: Thanks, I think this is an easy fix, also you already suggested the solutions:)

File: R/pubchem.R (hunk 2 above)
Comment (stitam) [reply R/pubchem.R:104]: For me it returns the correct response, but this is the example that is failing on both Appveyor and Travis. Very interesting.

File: R/pubchem.R (hunk 1 above)
Comment (stitam) [reply R/pubchem.R]: Updated the documentation.

File: R/pubchem.R (hunk 2 above)
Comment (stitam) [reply R/pubchem.R:102]: URL encoding the query fixed the Appveyor and Travis issue, thanks @Aariq!! FYI, encoding threw an error for non-character queries (e.g. CID) but that was also easy to fix.

**Pull request merged**